protocols/raft/src/main/java/io/atomix/protocols/raft/roles/LeaderRole.java

```diff
@@ -1368,51 +1362,71 @@ private void commitCommand(CommandRequest request, CompletableFuture<CommandResp
                                   .build());
                         }
                       });
-            },
-            raft.getThreadContext());
+            });
   }
 
   /**
-   * Appends an entry to the Raft log and compacts logs if necessary.
+   * Appends an entry to the Raft log.
    *
    * @param entry the entry to append
    * @param <E> the entry type
    * @return a completable future to be completed once the entry has been appended
    */
-  private <E extends RaftLogEntry> CompletableFuture<Indexed<E>> appendAndCompact(E entry) {
-    return appendAndCompact(entry, 0);
-  }
+  private <E extends RaftLogEntry> CompletableFuture<Indexed<E>> append(E entry) {
+    CompletableFuture<Indexed<E>> resultingFuture = null;
+    int retries = 0;
 
-  /**
-   * Appends an entry to the Raft log and compacts logs if necessary.
-   *
-   * @param entry the entry to append
-   * @param attempt the append attempt count
-   * @param <E> the entry type
-   * @return a completable future to be completed once the entry has been appended
-   */
-  protected <E extends RaftLogEntry> CompletableFuture<Indexed<E>> appendAndCompact(
-      E entry, int attempt) {
-    if (attempt == MAX_APPEND_ATTEMPTS) {
-      return Futures.exceptionalFuture(
-          new StorageException.OutOfDiskSpace("Not enough space to append entry"));
-    } else {
+    do {
       try {
-        return CompletableFuture.completedFuture(raft.getLogWriter().append(entry))
-            .thenApply(
-                indexed -> {
-                  log.trace("Appended {}", indexed);
-                  return indexed;
-                });
-      } catch (StorageException.TooLarge | BufferOverflowException e) {
-        return Futures.exceptionalFuture(e);
-      } catch (StorageException.OutOfDiskSpace e) {
-        log.warn("Caught OutOfDiskSpace error! Force compacting logs...");
-        return raft.getServiceManager()
-            .compact()
-            .thenCompose(v -> appendAndCompact(entry, attempt + 1));
+        resultingFuture = tryToAppend(entry);
+      } catch (StorageException storageException) {
+
+        // storage exception wraps IOException's
+        retries++;
+        if (retries > MAX_APPEND_ATTEMPTS) {
+          // only solution is to step down now
+          raft.transition(Role.FOLLOWER);
```

Comment:
There's no easy way to discard queued writes, is there? So they will still be executed, but we should guard against them by using the `isRunning` flag (or method? I forget, but there's something).

Reply:
Not 100% sure but it seems they are executed any way, but this is the same behavior as before. Eventually they are truncated from the log since the new leader has a different log. They are not replicated, since the appender is stopped.

Reply:
I found no way to clear the queue.

Reply:
Can we add a check using `isRunning`? See for example `safeAppendEntry` where we check that. If we're not running we shouldn't be appending at all.

Reply:
But then it is fine right, because we already have this check there? :sweat_smile:

Reply:
This only matters if we switch the context or append an task to the queue, which we only do on our appendEntry - on the others we directly execute the append method

Reply:
But that check is done when we enqueue, once it's queued we never check again. So if we don't check before writing, then we might be doing unnecessary I/O no?

Reply:
```